python/pyspark/rdd.py

```diff
@@ -704,7 +704,12 @@ def pipe_objs(out):
                     out.write(s.encode('utf-8'))
                 out.close()
             Thread(target=pipe_objs, args=[pipe.stdin]).start()
-            return (x.rstrip(b'\n').decode('utf-8') for x in iter(pipe.stdout.readline, b''))
+            result = (x.rstrip(b'\n').decode('utf-8') for x in iter(pipe.stdout.readline, b''))
+            pipe.wait()
```

Comment:
While waiting, the stdout of subprocess is not consumed, it may block the subprocess, causing dead lock.